homeassistant/components/syncthru/__init__.py

```diff
@@ -1,23 +1,71 @@
 """The syncthru component."""
 
+from typing import Set, Tuple
+
+from pysyncthru import SyncThru
+
 from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
 from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_URL
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import aiohttp_client, device_registry as dr
 from homeassistant.helpers.typing import ConfigType, HomeAssistantType
 
+from .const import DEFAULT_MODEL, DEFAULT_NAME_TEMPLATE, DOMAIN
+from .exceptions import SyncThruNotSupported
+
 
 async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
     """Set up."""
+    hass.data.setdefault(DOMAIN, {})
     return True
 
 
 async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
     """Set up config entry."""
+
+    session = aiohttp_client.async_get_clientsession(hass)
+    printer = hass.data[DOMAIN][entry.data[CONF_URL]] = SyncThru(
+        entry.data[CONF_URL], session
+    )
+
+    try:
+        await printer.update()
+    except ValueError as ex:
+        raise SyncThruNotSupported from ex
+    else:
+        if printer.is_unknown_state():
+            raise ConfigEntryNotReady
+
+    device_registry = await dr.async_get_registry(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        connections=device_connections(printer),
+        identifiers={(DOMAIN, printer.serial_number())},
+        model=printer.model(),
+        name=DEFAULT_NAME_TEMPLATE.format(printer.model() or DEFAULT_MODEL),
```

Comment:
Feels weird to inject the model into the name since there is also a model field.

Reply:
Changed to use printer's "hostname"; that's the closest thing for a name here and also used to some extent by the ipp integration for these devices.